feat(curriculum): add string method workshop (#66936)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/blocks/workshop-employee-profile-validator/69e20876baed33de636b989a.md

---
id: 69e20876baed33de636b989a
title: Step 3
challengeType: 20
dashedName: step-3
---

# --description--

Sometimes, you may want to find the position of a specific character inside a string. Python provides the built-in `.find()` method for this purpose. It returns an integer representing the index (position) of the first occurrence of the given substring. For example:

```py
text = 'hello'
print(text.find('e'))  # 1
```

> Now, call the `.find()` method on `email` and pass it the string `'@'` to find the position of the at sign in the `email` string. Assign the result to a variable named `at_position`.

Reviewer: ```suggestion
Now, call the `.find()` method on `email` and pass it `'@'` to find the position of the at sign in the `email` string. Assign the result to a variable named `at_position`.
```

> Then, print the result of concatenating the string `Position of @: ` with the value of `at_position`. Remember to use `str()` to convert `at_position` before concatenating it.

Reviewer 2: Steps 5 onward write string literals with their quotes inside the backticks. Making it consistent here.

```suggestion
Now, call the `.find()` method on `email` and pass it the string `'@'` to find the position of the at sign in the `email` string. Assign the result to a variable named `at_position`.

Then, print the result of concatenating the string `'Position of @: '` with the value of `at_position`. Remember to use `str()` to convert `at_position` before concatenating it.
```

> Then, print the result of concatenating the string `'Position of @: '` with the value of `at_position`. Remember to use `str()` to convert `at_position` before concatenating it.

Reviewer: ```suggestion
Then, print the result of concatenating the string `Position of @: ` with the value of `at_position`. Remember to use `str()` to convert `at_position` before concatenating it.
```

# --hints--

You should create a variable named `at_position`.

```js
({ test: () => assert(runPython(`_Node(_code).has_variable('at_position')`)) })
```

You should assign `at_position` the value `email.find('@')`.

```js
({ test: () => assert(runPython(`
_Node(_code).find_variable('at_position').is_equivalent("at_position = email.find('@')")
`)) })
```

> You should print the result of concatenating the string `Position of @: ` with the `at_position` variable.

Reviewer 2: ```suggestion
You should print the result of concatenating the string `'Position of @: '` with the `at_position` variable.
```

> You should print the result of concatenating the string `'Position of @: '` with the `at_position` variable.

Reviewer: ```suggestion
You should print the result of concatenating the string `Position of @: ` with the `at_position` variable.
```

---
id: 69e20876baed33de636b989a
title: Step 2
challengeType: 20
dashedName: step-2
---

> 

Reviewer: As per a previous comment, this step should **only** print `email`. If you check, the current step 2 has the test for it but not the instructions to actually do it.

# --description--

> Sometimes, you may want to find the position of a specific character inside a string. Python provides the built-in `.find()` method for this purpose. It returns an integer representing the index (position) of the first occurrence of the given substring. 

Reviewer 2: Trailing whitespace, and this is the only line in the workshop that has it.

```suggestion
Sometimes, you may want to find the position of a specific character inside a string. Python provides the built-in `.find()` method for this purpose. It returns an integer representing the index (position) of the first occurrence of the given substring.
```

```py
text = 'hello'
print(text.find('e'))  # 1
```

> Now, call the `.find()` method on `email` and pass it the string `'@'` to find the position of the at sign in the `email` string. Assign the result to a variable named `at_position`.

Reviewer: ```suggestion
Now, call the `.find()` method on `email` and pass it the string `@` to find the position of the at sign in the `email` string. Assign the result to a variable named `at_position`.
```

Now, call the `.find()` method on `email` and pass it `'@'` to find the position of the at sign in the `email` string. Assign the result to a variable named `at_position`.

> Then print `at_position` so that the output reads: `Position of @: 13`

Reviewer 2: L19 calls the character "the at sign", while step 4 calls it "the `@` sign" and step 12 "the `@` character", so it is named three ways in prose. The suggestion adopts step 12's form. The printed labels ("Has @ symbol") are output text and should stay as they are.

The trailing punctuation is the cross-step item: "so that the output reads" is followed by a colon and no period in steps 2, 3 and 5, a colon and a period in steps 6, 11, 13 and 16, and no colon with a period in steps 4, 9 and 15. Every hint line in the workshop uses the last of those, so aligning the descriptions to it is the smallest change.

```suggestion
Now, call the `.find()` method on `email` and pass it `'@'` to find the position of the `@` character in the `email` string. Assign the result to a variable named `at_position`.

Then print `at_position` so that the output reads `Position of @: 13`.
```

> Then, print the result of concatenating the string `'Position of @: '` with the value of `at_position`. Remember to use `str()` to convert `at_position` before concatenating it.

Reviewer: ```suggestion
Then, print the result of concatenating the string `Position of @: ` with the value of `at_position`. Remember to use `str()` to convert `at_position` before concatenating it.
```

> Then print the `email` with the prefix `'Email: '` and print `at_position` with the prefix `'Position of @: '`. Remember to use `str()` to convert `at_position` to properly concatenate it with the prefix.

Reviewer 3: ```suggestion
Now, call the `.find()` method on `email` and pass it the string `'@'` to find the position of the at sign in the `email` string. Assign the result to a variable named `at_position`.

Then, print the result of concatenating the string `'Position of @: '` with the value of `at_position`. Remember to use `str()` to convert `at_position` before concatenating it.
```

# --hints--

You should create a variable named `at_position`.

```js
({ test: () => assert(runPython(`_Node(_code).has_variable('at_position')`)) })
```

> You should assign `at_position` the value `email.find('@')`.

Reviewer 2: The hints that grade an assignment use five different phrasings for the same thing:

| Form | Steps |
| --- | --- |
| "You should assign X to `var`" | 1, 7, 10, 14, 15, 17 |
| "You should assign `var` the value X" | 3, 16 |
| "Your `var` variable should have the value X" | 5, 6 |
| "You should use M on X to assign `var` a value" | 4, 8, 20 |
| "You should call/use M on X and assign the result to the `var` variable" | 10, 11, 12, 13, 14, 15, 17, 18 |

Two of these cover every case: "You should assign X to `var`." for a literal or plain expression, and "You should call `.method()` on `x` and assign the result to the `var` variable." for a method call, which is already the largest group.

This step is the clearest instance, since step 12 grades the identical `.find('@')` call and words it the way the suggestion does. The rest: step 4 L31, step 8 L28, step 20 L27 and L42 to the method-call form; step 5 L30, step 6 L22 and step 16 L30 to the assign form.

```suggestion
You should call `.find('@')` on `email` and assign the result to the `at_position` variable.
```

({ test: () => assert(runPython(`
node = _Node(_code).find_variable('at_position')

> not node.is_empty() and node.is_equivalent("at_position = email.find('@')")

Reviewer: ```suggestion
_Node(_code).find_variable('at_position').is_equivalent("at_position = email.find('@')")
```

`)) })
```

> You should print `email` with the prefix `'Email: '`.

Reviewer: ```suggestion
You should print the result of concatenating the string `Email: ` with the `email` variable.
```

```js
({ test: () => assert(runPython(`_Node(_code).has_call("print('Email: ' + email)")`)) })
```

> You should print `at_position` with the prefix `'Position of @: '`.

Reviewer: ```suggestion
You should print the result of concatenating the string `Position of @: ` with the `at_position` variable.
```

Now, find the position of the `'@'` symbol in the `email` string.

Create a variable named `at_position` and assign it the result of calling `.find()` on `email`. 

Then, print the `email` variable with the label `'Email: '`.

Also, print the value of `at_position` with the label `'Position of @: '`. Remember to use `str()` to convert `at_position` before concatenating it with the label.

# --hints--

You should create a variable named `at_position`.

```js
({ test: () => assert(runPython(`_Node(_code).has_variable('at_position')`)) })
```

You should assign `at_position` using `email.find('@')`.

```js
({ test: () => assert(runPython(`'email.find' in _code and 'at_position' in _code`)) })
```

You should print the `email` with the label `'Email: '`.

```js
({ test: () => assert(runPython(`'Email: ' in _code and 'email' in _code`)) })
```

You should print the `position` with the label `'Position of @: '`.

```js
({ test: () => assert(runPython(`'Position of @: ' in _code and 'at_position' in _code`)) })
```

# --seed--

## --seed-contents--

> ```py
> email = 'alice.johnson@company.com'
> --fcc-editable-region--

Reviewer 3: ```suggestion
email = 'alice.johnson@company.com'

--fcc-editable-region--
```
